File: code/3sum_closest.java
public class Solution {
    /*
     * @param numbers: Give an array numbers of n integer
     * @param target: An integer
     * @return: return the sum of the three integers, the sum closest target.
     */
    public int threeSumClosest(int[] numbers, int target) {
        // write your code here

        Arrays.sort(numbers);
        int diff = Integer.MAX_VALUE;
        int ans = 0;
        for (int i = 0; i < numbers.length; i++) {
            rest = target - numbers[i];
            int start = i + 1;
            int end = numbers.length - 1;
            while (start < end) {
                int sum = numbers[start] + numbers[end];
                int tempDiff = Math.abs(rest - sum);
                if (tempDiff < diff) {
                    diff = tempDiff;
                    ans = sum + numbers[i];
                    if (tempDiff == 0) return ans;
                }
                if (sum > rest) {
                    end--;
                } else {
                    start++;
                }
            }
        }
        return ans;

    }

}

File: code/medium1/3sum_closest.cpp
// 不要搞中间突然break，不对
// 比如：{-55,-24,-18,-11,-7,-3,4,5,6,9,11,23,33}
// 因为vector中及有正，又有负，加出来以后sum不一定是单调变化的
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {

        sort(nums.begin(), nums.end());
        
        int minDiff = INT_MAX;
        int ans = 0;
        for (int i = 0; i < nums.size(); i++) {
            int tempTarget = target - nums[i];
            int start = i + 1, end = nums.size() - 1;
            int sum = 0;
            int diff = INT_MAX;
            while (start < end) {
                sum = nums[start] + nums[end];
                if (sum < tempTarget) {
                    // if (tempTarget - sum > diff) break;
                    diff = tempTarget - sum;
                    if (diff <= minDiff) {
                        ans = sum + nums[i];
                        minDiff = tempTarget - sum;
                    }
                    start++;
                } else if (sum > tempTarget) {
                    // if (sum - tempTarget > diff) break;
                    diff = sum - tempTarget;
                    if (sum - tempTarget <= minDiff) {
                        ans = sum + nums[i];
                        minDiff = sum - tempTarget;
                    }
                    end--;
                } else {
                    return target;
                }
            }
        }
        return ans;
    }
};

// the lower bound for 3sum is O(n2)